src/test/java/net/minecraftforge/debug/network/PacketTest.java

```diff
@@ -0,0 +1,124 @@
+package net.minecraftforge.debug.network;
+
+import org.slf4j.Logger;
+
+import com.mojang.logging.LogUtils;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.gametest.framework.GameTest;
+import net.minecraft.gametest.framework.GameTestHelper;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.level.GameType;
+import net.minecraft.world.level.Level;
+import net.minecraftforge.event.network.CustomPayloadEvent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
+import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
+import net.minecraftforge.gametest.GameTestHolder;
+import net.minecraftforge.network.Channel.VersionTest;
+import net.minecraftforge.network.ChannelBuilder;
+import net.minecraftforge.network.PacketDistributor;
+import net.minecraftforge.network.SimpleChannel;
+import net.minecraftforge.test.BaseTestMod;
+
+@GameTestHolder("forge." + PacketTest.MODID)
+@Mod(PacketTest.MODID)
+public class PacketTest extends BaseTestMod {
+    static final String MODID = "packet";
+    
+    protected static final Logger LOGGER = LogUtils.getLogger();
+
+    public PacketTest(FMLJavaModLoadingContext context) {
+        super(context);
+    }
+    
+    @SubscribeEvent
+    public void commonSetup(final FMLCommonSetupEvent event) {
+        PacketHandler.registerMessages();
+    }
+    
+    @GameTest(template = "forge:empty3x3x3")
+    public static void send_to_player(GameTestHelper helper) {
+        // Send a test packet directly to a player
+        ServerPlayer player = helper.makeMockServerPlayer();
+        PacketHandler.sendToPlayer(new TestPacket("testing123"), player);
+        helper.succeed();
+    }
+    
+    @GameTest(template = "forge:empty3x3x3")
+    public static void send_to_nearby(GameTestHelper helper) {
+        // Send a test packet to all players near a target point
+        helper.makeMockPlayer(GameType.SURVIVAL);
+        PacketHandler.sendToAllAround(new TestPacket("testing123"), Level.OVERWORLD, BlockPos.ZERO, 5D);
+        helper.succeed();
```

Comment:
Is there any way to tell that the clients have received these packets? perhaps adding a few mock players to the server and then counting all the received packets?

Reply:
I'm sorry, but I can't think of a way to sanely do what you're asking. To my knowledge, the Channel object provides no mechanism for tracking who's received what packets, even in a test environment. I could *maybe* send a reply packet that modifies the server player in some detectable way, but that feels really clunky to me. Apologies if I'm misunderstanding what you're asking for; I haven't been operating at my best for the last week.